File: BenchmarkRunner/Benchmarks/C#/Binary Trees.cs
using CsharpRAPL.Benchmarking.Attributes;
using CsharpRAPL.Benchmarking.Attributes.Parameters;
using CsharpRAPL.Benchmarking.Lifecycles;

namespace BenchmarkRunner.Benchmarks.C_;

public class BinaryTrees
{
    const int MinDepth = 4;
    const int NoTasks = 1;
    
    [Benchmark("Binary Trees","Binary trees in C#", name:"C sharp BT", skip: false)]
    public static int BinaryTree([BenchmarkLoopiterations] ulong LoopIterations)
    {
        int maxDepth = 8;
        int result = 0;

        //Console.WriteLine(string.Concat("stretch tree of depth ", maxDepth + 1,"\t check: ", (bottomUpTree(maxDepth + 1)).itemCheck()));
        var longLivedTree = bottomUpTree(maxDepth);
        for (ulong i = 0; i < LoopIterations; i++)
        {
            

            //var results = new string[(maxDepth - MinDepth) / 2 + 1];
            int depth = maxDepth;
            int n = (1 << depth + MinDepth) / NoTasks;
            var tasks = new int[NoTasks];
            for (int t = 0; t < tasks.Length; t++)
            {
                var checkt = 0;
                for (int it = n; it > 0; it--)
                    checkt += (bottomUpTree(depth)).itemCheck();
                tasks[t] = checkt;
            } 
            var check = tasks[0];
            for (int t = 1; t < tasks.Length; t++)
                check += tasks[t];
            result += check;
        }
        
            //results[i] = string.Concat(n * NoTasks, "\t trees of depth ",
              //  depth, "\t check: ", check);
              //for (int i = 0; i < results.Length; i++)
            //Console.WriteLine(results[i]);

        //Console.WriteLine(string.Concat("long lived tree of depth ", maxDepth,"\t check: ", longLivedTree.itemCheck()));
        return result;
    }
    
    static TreeNode bottomUpTree(int depth) 
    {
        if (0 < depth) {
            return new TreeNode(bottomUpTree(depth - 1), bottomUpTree(depth - 1));
        }
        return new TreeNode(null,null);
    }

    private class TreeNode
    {
        readonly TreeNode? left, right;
        
        internal TreeNode(TreeNode left, TreeNode right) 
        {
            this.left = left;
            this.right = right;
        }     

        internal int itemCheck()
        {
            if (left == null) {
                return 1;
            }
            return 1 + left.itemCheck() + right.itemCheck();
        }
    }   
}

File: BenchmarkRunner/Benchmarks/Java/BinaryTrees.java
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
public class BinaryTrees {

    public static long binaryTrees(long loopIterations) throws Exception {
        int n = 21;
        int minDepth = 4;
        int maxDepth = Math.max(minDepth + 2, n);
        int stretchDepth = maxDepth + 1;
        int check = (TreeNode.create(stretchDepth)).check();

        //System.out.println("stretch tree of depth " + (maxDepth + 1) + "\t check: " + check);

        TreeNode longLivedTree = TreeNode.create(maxDepth);
        long result = 0;
        for(long l = 0; l < loopIterations; l++){
            for (int depth = minDepth; depth <= maxDepth; depth += 2)
            {
                int iterations = 1 << (maxDepth - depth + minDepth);
                check = 0;

                for (int i = 1; i <= iterations; i++)
                {
                    check += (TreeNode.create(depth)).check();
                }
                //System.out.println(iterations + "\t trees of depth " + depth + "\t check: " + check);
                result += check;
            }

        }

        return result;
        //System.out.println("long lived tree of depth " + maxDepth + "\t check: " + longLivedTree.check());
    }

    static class TreeNode {
        TreeNode left, right;

        static TreeNode create(int depth)
        {
            return ChildTreeNodes(depth);
        }

        static TreeNode ChildTreeNodes(int depth)
        {
            TreeNode node = new TreeNode();
            if (depth > 0)
            {
                node.left = ChildTreeNodes(depth - 1);
                node.right = ChildTreeNodes(depth - 1);
            }
            return node;
        }

        int check() {
            return left == null ? 1 : left.check() + right.check() + 1;
        }
    }
}
